extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java

```diff
@@ -830,12 +901,72 @@ private String getEmbeddedInfoStsAndStorePlayerJsUrl() {
                 }
             }
 
-            // Get embed sts
-            return Parser.matchGroup1("\"sts\"\\s*:\\s*(\\d+)", embedPageContent);
-        } catch (final Exception i) {
-            // if it fails we simply reply with no sts as then it does not seem to be necessary
-            return "";
+            if (playerJsUrl != null) {
+                if (playerJsUrl.startsWith("//")) {
+                    playerJsUrl = HTTPS + playerJsUrl;
+                } else if (playerJsUrl.startsWith("/")) {
+                    // Sometimes https://www.youtube.com part has to be added manually
+                    playerJsUrl = HTTPS + "//www.youtube.com" + playerJsUrl;
+                }
+                playerCode = NewPipe.getDownloader().get(playerJsUrl, getExtractorLocalization())
+                        .responseBody();
+            } else {
+                throw new ExtractionException("Could not extract JS player URL");
+            }
+        } catch (final Exception e) {
+            throw new ParsingException("Could not store JavaScript player", e);
+        }
+    }
+
+    private boolean hasOtfStreams() {
+        if (streamingData != null) {
+            boolean hasOtfStreamsValue = false;
+            if (streamingData.has("adaptiveFormats")) {
+                final JsonArray adaptiveFormats = streamingData.getArray("adaptiveFormats");
+                for (final Object adaptiveFormat : adaptiveFormats) {
+                    final JsonObject jsonAdaptiveFormat = (JsonObject) adaptiveFormat;
+                    if (jsonAdaptiveFormat.has("type")) {
+                        final String streamTypeFormat = jsonAdaptiveFormat.getString("type",
+                                EMPTY_STRING);
```

Comment:
You can refactor all of that.

You don't need to check if it has `adaptiveFormats` because getArray() by default returns an empty array.
Typical use case for `.has("key")` function would do an if (has("key")) {do something} else { do something else}, but here you don't do anything if it doesn't have the key, so it's not needed to check

same for has("type"), if it doesn't have it will return an EMPTY_STRING as you explicitly told, and "" is not equals to "FORMAT_STREAM_TYPE_OTF" anyway